### PR TITLE
[WIP] Added support for doctrine annotations

### PR DESCRIPTION
**Pull request by core23**
I am targeting this branch, because {reason}.

## Changelog

```markdown
### Added
- Added support for doctrine annotations
```

    
## To do
    
- [ ] Update the tests
- [ ] Add an upgrade note
- [ ] Test it on a real project

## Subject

The DiExtraBundle has no support for symfony 4 and looks abandoned. This PR will use `doctrine/annotations` to add support for all symfony versions.

### CODE REVIEW

File: src/Annotation/Admin.php [hunk 1]
```diff
@@ -173,3 +178,22 @@ private function generateFallback($name)
         }
     }
 }
+
+// NEXT_MAJOR: Remove this hack together with DiExtraBundle
```
Comment (kunicmarko20) [inline src/Annotation/Admin.php]: you should add deprecations

File: src/DependencyInjection/Compiler/AutoRegisterCompilerPass.php [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+use Doctrine\Common\Annotations\AnnotationReader;
+use Sonata\AdminBundle\Annotation\Admin;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\Finder\Finder;
+use Symfony\Component\Finder\SplFileInfo;
+
+final class AutoRegisterCompilerPass implements CompilerPassInterface
```
Comment (kunicmarko20) [inline src/DependencyInjection/Compiler/AutoRegisterCompilerPass.php]: code here looks familiar :smile:

File: src/DependencyInjection/SonataAdminExtension.php [hunk 3]
```diff
@@ -454,6 +419,49 @@ public function getNamespace()
...
+                    return;
+                }
+                $annotationPatternsConfigured = true;
+
+                break;
+            }
+        }
+
+        @trigger_error(
+            'Automatic registration of annotations is deprecated since 3.14, to be removed in 4.0.',
```
Comment (kunicmarko20) [inline src/DependencyInjection/SonataAdminExtension.php]: maybe now mention that doctrine should be used

File: docs/reference/annotations_doctrine.rst [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    /**
+     * @Sonata\Admin(
+     *   class="AcmeBundle\Entity\MyEntity",
+     *   id="service id (generated per default)",
+     *   managerType="doctrine_mongodb (orm per default)",
+     *   baseControllerName="Sonata\AdminBundle\Controller\CRUDController",
+     *   group="myGroup",
+     *   label="myLabel",
+     *   showInDashboard=true,
+     *   translationDomain="AppBundle",
```
Comment (OskarStark) [inline docs/reference/annotations_doctrine.rst]: This should match the namespace of the example

File: docs/reference/annotations_doctrine.rst [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+     *   pagerType="",
+     *   persistFilters="",
```
Comment (OskarStark) [inline docs/reference/annotations_doctrine.rst]: Empty string? No true/false?

File: docs/reference/annotations_doctrine.rst [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+     *   icon="<i class='fa fa-folder'></i>",
+     *   keepOpen=false,
+     *   onTop=false
+     * )
+     */
+    class MyAdmin extends AbstractAdmin
+    {
```
Comment (OskarStark) [inline docs/reference/annotations_doctrine.rst]: Please add „// ...“ to make clear this class is not empty

File: docs/reference/annotations_doctrine.rst [hunk 7]
```diff
@@ -0,0 +1,55 @@
+Annotations using doctrine
+==========================
+
+All annotations require doctrine/annotations, it can easily be installed by composer:
```
Comment (OskarStark) [inline docs/reference/annotations_doctrine.rst]: Please remove „easily“

File: src/DependencyInjection/Compiler/AutoRegisterCompilerPass.php [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+{
+    const DEFAULT_SERVICE_PREFIX = 'app.admin.';
```
Comment (OskarStark) [inline src/DependencyInjection/Compiler/AutoRegisterCompilerPass.php]: Do we Need this const?

File: src/DependencyInjection/Compiler/AutoRegisterCompilerPass.php [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    /**
+     * @param string $directory
+     *
+     * @return \IteratorAggregate
+     */
+    private function findFiles($directory)
+    {
+        return Finder::create()
```
Comment (OskarStark) [inline src/DependencyInjection/Compiler/AutoRegisterCompilerPass.php]: This should be an explicit dependency of our bundle if not already

**Comment by OskarStark** [comment]
Any news here @core23 ?

File: src/DependencyInjection/Compiler/AutoRegisterCompilerPass.php [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+            return;
+        }
+
+        $reader = $container->get('annotation_reader');
+
+        foreach ($this->findFiles($container->getParameter('sonata.admin.annotations.directory')) as $file) {
+            if (!($className = $this->getFullyQualifiedClassName($file))) {
+                continue;
+            }
+
```
Comment (kunicmarko20) [inline src/DependencyInjection/Compiler/AutoRegisterCompilerPass.php]: Here you need https://github.com/kunicmarko20/SonataAnnotationBundle/pull/24 this check

File: src/DependencyInjection/Compiler/AutoRegisterCompilerPass.php (hunk 8 above)
Comment (core23) [reply src/DependencyInjection/Compiler/AutoRegisterCompilerPass.php]: There is one usage for it, so yes

**Comment by core23** [comment]
Closing this in favor of @kunicmarko20 bundles

**Pull request closed without merging**